### PR TITLE
Initialize CloudProviders with a SecretKeyGetter

### PR DESCRIPTION
**Pull request by alvaroaleman**
**What this PR does / why we need it**:

This PR changes the cloud provider interface in a away that they get initialized with a SecretKeyGetter, ensuring they are always able to get credentials. This fixes various AWS endpoints.

The whole thing is done by:

* Creating a factory for a `SecretKeySelectorValueFunc` in the provider pkg
* Asserting the `ClusterProvider` to its implementation `*kubernetesprovider.ClusterProvider` then using its `GetSeedClusterAdminRuntimeClient` to call the factory
* For all endpoints that do not have a cluster already, the `ClusterProvider` can not be uses as its creation needs a DC which we don't have. Hence a `SeedClientGetter` was added to the `Routing`, it is an abstraction on top of the `KubeconfigGetter` with an integrated restMapper cache. It needs the seed thought, which is why `DatacenterFromSeedMap` had to be extended to also return the seed

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

What is really bad about all of this is that a lot of changes had to be done in order to have a `SecretKeySelector` in endpoints that do not operate on existing clusters _even thought it will never be used_ down the line in the AWS provider.

IMHO we should refactor the VPC/Subnet/Zone endpoints to not require an initialized AWS cloud provider but instead be standalone and just take the credentials as args.

What I will do as follow-ups are:
* DRY out restmapper cache usage and use the `*restmapper.Cache` everywhere
* DRY out `SecretKeySelectorValueFunc` creation and use the factory in the provider package everywhere

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
none
```

/assign @mrIncompetent @zreigz @thz

### CODE REVIEW

File: api/pkg/util/restmapper/cache.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/client-go/rest"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// New returns a new Cache
+func New() *Cache {
```
Comment (mrIncompetent) [inline api/pkg/util/restmapper/cache.go:13]: This should also be used in `api/pkg/cluster/client/client.go` otherwise we have 2 different cache mechanisms

Comment (alvaroaleman) [reply api/pkg/util/restmapper/cache.go:13]: Yeah I mentioned in the PR description that I want to DRY out the restmapper caching everywhere, but as a follow-up, to unbreak master asap

Comment (mrIncompetent) [reply api/pkg/util/restmapper/cache.go:13]: Sorry, missed reading the description :man_facepalming:

**Pull request merged**